libraries/standard/mqtt/src/mqtt.c

```diff
@@ -601,30 +818,124 @@ static MQTTStatus_t handleIncomingPublish( MQTTContext_t * const pContext,
     LogInfo( ( "De-serialized incoming PUBLISH packet: DeserializerResult=%d", status ) );
 
     if( status == MQTTSuccess )
+    {
+        /* Check if dup bit is set by the broker. Duplicates are not possible
+         * for QoS0 publishes. */
+        if( ( publishInfo.dup ) && ( ( publishInfo.qos == MQTTQoS1 ) ||
+                                     ( publishInfo.qos == MQTTQoS2 ) ) )
+        {
+            /* Check if the publish packet was already received. */
+            hasPublishReceivedBefore = checkIfPublishWasReceivedBefore( pContext,
+                                                                        packetIdentifier,
+                                                                        publishInfo.qos,
+                                                                        &isStateUpdateNeeded,
+                                                                        &publishRecordState );
+        }
+    }
+
+    /* State update for receive only for packets which are not duplicate
+     * incoming publishes. */
+    if( ( status == MQTTSuccess ) && ( !hasPublishReceivedBefore ) )
     {
         publishRecordState = MQTT_UpdateStatePublish( pContext,
                                                       packetIdentifier,
                                                       MQTT_RECEIVE,
                                                       publishInfo.qos );
         LogInfo( ( "State record updated. New state=%d.",
                    publishRecordState ) );
+    }
 
+    if( status == MQTTSuccess )
+    {
+        /* Pass the incoming publish to application, except the QoS2 duplicate
+         * incoming publishes. */
+        if( ( !hasPublishReceivedBefore ) || ( publishInfo.qos != MQTTQoS2 ) )
```

Comment:
Why not QoS1?

Reply:
QoS1 will always be passed to application as it allows duplicates(at least once delivery). On the other hand, library will not pass any duplicate QoS2 publishes to application(at most once).

Reply:
We should specify that in a comment